settings.php

```
@@ -52,6 +52,19 @@ function ecs($configoption, $params = null) {
 
     $settings->add(new admin_setting_configtext('equella_default_window_height', ecs('window.height'), '', EQUELLA_DEFAULT_WINDOW_HEIGHT));
 
+    // ///////////////////////////////////////////////////////////////////////////////
+    //
+    // SSO settings
+    //
+    $settings->add(new admin_setting_heading('equella_sso_settings', ecs('sso.heading'), ''));
+
+    $userfieldoptions = [];
+    $userfieldoptions = ['default' => ecs('userfield.username')];
```

Can we use username here instead of default for the key?

---

We can, but there is a risk that one of the custom profile fields would be named as "username" (for some reason). That would mess everything. So we took approach from https://tracker.moodle.org/browse/MDL-79959 and applied it here for consistency.

---

Ok. But if that's the case then what if custom profile fields is named as "default" as shortname. What happens in that case? I assume that will also mess things up.

---

Fair enough. 
I'll make this change.

---

Hey @edalex-siddharth 
I was going to implement this in suggested way, but then I realised that this would add an extra SQL call for all users. We run this on a very busy instance and would avoid any extra SQLs if possible. 

So I took a bit different approach:
1. Abstracted fields fields logic into a new \mod_equella\user_field class. 
2. Add username, idnumber and email support as well as all text custom profile fields.
3. Now I prefix all custom profile fields in the settings so there is no way it can clash with standard user fields like username or email.
4. Finally I modified mod_equella_get_userfield_value function to glue it all together. 

I believe this is flexible and extendable approach. Hopefully you are ok with that. 

The only thing I'm not sure about is having a fallback in case if configured field is empty for a user. IMHO it should throw an exception and explode preventing any possible clashes with usernames. So potentially a value for a different user can be used as userfield. E.g. when custom field is empty and user name is equal a value of custom field of a different user.

---

Hi @dmitriim , can you please resolve the merge conflicts before I review the code?

---

This has been done.

---

Hi @dmitriim,

Instead of a fallback, can we do either of the following:

1. Filter out the fields which are empty and only allow fields which are not empty and have some value. That way this fallback won't be necessary and if `username` is the only option to select from then we can just disable the select input.
2. Instead of filtering out maybe we can show a warning/error text beneat the select input showing some message like, this field is empty and should have some valid value, something like this and prevent the user from selecting the userfield which might be empty. 

IMO first option seems like a good alternative to avoid this problem. I am curious to know what you think?